feat: JWT 구현, OAuth2 구현 및 Security Configuration 마무리

wan2daaa opened the pull request:
## JWT
accessToken 과 refreshToken 으로 나눠서 구현하였고,
refreshToken은 db 에 저장하고,
accessToken 은 필터(TokenAuthenticationFilter.java)에서
검증 되면 정상적으로 로직 진행
검증 실패시 -> refreshToken을 보고 유효할 시 , accessToken 재발급
->
accessToken 은 RequestHeader Authorization : Bearer {jwt}
refreshToken 은 Cookie (Http Only) refresh_token : {jwt} 로 삽입


## OAuth2
Oauth2 로그인 성공 후
refreshToken 과 accessToken을 전달하는 방법은 위 로그인 과정과 동일 
-> 
accessToken 은 response Header에 Authorization: Bearer {jwt}
refreshToken은 cookie (Http Only) refresh_token : {jwt} 로 삽입

로그아웃 시refreshToken 삭제 로직 추가

# 체크해주셨으면 하는 부분 
Outh2 쪽이 조금 부족한 것같아서 그쪽 위주로 체크해 주시면 감사하곘습니다!

wan2daaa committed:
feat: jwt issuer, secret_key 추가

wan2daaa committed:
feat: 테스트 패키지에서도 lombok 사용할 수 있게 추가

wan2daaa committed:
feat: 테스트 패키지가 존재하지 않아서 기본 테스트 클래스 생성 후 테스트

wan2daaa committed:
feat: jwt 구현시 User 엔티티가 필요하여 생성 및 이에 따른 클래스도 생성

wan2daaa committed:
feat: jwt 관련 issuer, secret_key 를 클래스로 사용하게 구현

wan2daaa committed:
feat: jwt 구현 (accessToken ,refreshToken)

accessToken 과 refreshToken 으로 나눠서 구현하였고,
refreshToken은 db 에 저장하고,
accessToken 은 필터(TokenAuthenticationFilter.java)에서
검증 되고 처리되고,

refreshToken은 POST /api/tokens 로 재발급 가능하고, 검증되고 accessToken을 리턴하게 구현

wan2daaa committed:
feat: 쿠키 유틸 구현 (OAuth2 시 필요)

wan2daaa committed:
feat: 스프링 시큐리티 Config 작성

csrf, cors, httpBasic, form, session
다 사용하지 않아서 막아 두었습니다

filterChain 이후는

사용할 시큐리티 관련 구현체에 DI해주는 코드 입니다

wan2daaa committed:
feat: OAuth2 코드 구현

Oauth2 로그인 성공 후
refreshToken 과 accessToken을 전달할 때
->
refreshToken은 cookie로
accessToken 은 url 파라미터로 전달하게 구현

로그아웃은 클라이언트 쪽에서 삭제하면 되어서 따로 처리하지 않았습니다.

5tr1ker commented on src/main/java/com/team/heyyo/config/SpringSecurityConfig.java:
```diff
@@ -1,46 +1,95 @@
+
     /* 비밀번호 암호화 */
     @Bean
     public PasswordEncoder passwordEncoder() {
```
RefreshToken은 서버에서 보관되기 때문에 클라이언트가 삭제하지 못하며 ,  서버에서 삭제해 주어야 합니다.

.logout()
.logoutUrl("/logout")
.deleteCookies("refreshToken")
를 사용해서 제거할 수 있습니다. ( /logout url로 들어가야해용

wan2daaa replied:
추가했습니다 감사합니다!

5tr1ker commented on src/main/java/com/team/heyyo/config/SpringSecurityConfig.java:
```diff
@@ -1,46 +1,95 @@
+
     /* 비밀번호 암호화 */
     @Bean
     public PasswordEncoder passwordEncoder() {
```
/api/tokens 도 좋지만 , 만료된 즉시 AccessToken 을 반환해서 api 통신 횟수를 줄일 수 있는 방법이 있습니다!

5tr1ker commented on src/main/java/com/team/heyyo/user/service/UserService.java:
```diff
@@ -0,0 +1,43 @@
+     * @return
+     */
+    public Long save(){
+        BCryptPasswordEncoder encoder = new BCryptPasswordEncoder();
```
BCryptPasswordEncoder 는 Configuration 에서 스프링 컨테이너에 보관되어있기 때문에 Dependency Injection 으로 주입할 수 있지 않을까요

wan2daaa replied:
맞습니다,,, 스프링 공부의 중요성

5tr1ker commented on src/main/java/com/team/heyyo/user/service/UserService.java:
```diff
@@ -0,0 +1,43 @@
+    public User findByEmail(String email) {
+        return userRepository.findByEmail(email)
+                .orElseThrow(() -> new IllegalArgumentException("Unexpected User"));
+    }
```
예외가 발생했을 때 이것을 담을 Exception Handler가 없는거같습니다! 그리구 메세지도 직관적으로 변경되면 좋을꺼같아요!

wan2daaa replied:
일단 임시로 만들어 놓았는데 그럼 구현하도록하겠습니다!

5tr1ker commented on src/main/java/com/team/heyyo/config/jwt/controller/TokenApiController.java:
```diff
@@ -0,0 +1,28 @@
+
+        return ResponseEntity.status(HttpStatus.CREATED)
+                .body(newAccessToken);
+    }
```
201은 새로운 자원이 생성되었을 때 사용되는데 AccessToken 은 자원 생성과는 거리가 있어보이는데 어떻게 생각하시나용

그리구 따로 변수로 빼지 않고 메서드를 바로 호출할 수 있을 꺼 같아요,
ResponseEntity.ok().body(tokenService.createNewAccessToken(request));

wan2daaa replied:
음 제 생각은 POST 요청으로 요청을 했으니까 이에 따른 자원 생성이 문맥상 맞다고 저는 생각하는데 그럼 혹시 어떤 걸로 수정하는게 좋을까요?? 

너무 인라인을 하게 되면 조금 가독성이 떨어져 보일까봐 구분하긴했는데 그럼 말씀하신것처럼 수정해보겠습니다!

wan2daaa committed:
feat: PasswordEncoderConfig 작성

SpringSecurityConfig 에 비밀번호 암호화 관련 코드 작성 시 순환 참조 하는 오류가 있어서 따로 분리

wan2daaa committed:
feat: 쿠키Util 에 HttpOnly로 만드는 메서드 추가

wan2daaa committed:
feat: 토큰 별 만료 기간을 enum 으로 구현

wan2daaa committed:
feat: accessToken 을 queryParam 에 담는 방식에서 header 에 담는 방식으로 변경

wan2daaa committed:
feat: application.yaml 에 frontend.url 경로 설정

wan2daaa committed:
feat: 토큰 유효성 검사 메소드에 토큰 만료되었는지 체크하는 로직 추가

wan2daaa committed:
refact: BCryptPasswordEncoder PasswordEncoder 인터페이스로 들고오게 설정

@Configuration 파일에 DI 를 해줬는데 실수로 구현체를 주입 해서 수정하였습니다

wan2daaa committed:
refact: 클래스 상수에서 JwtTokenDuration로 수정

wan2daaa committed:
refact: 코드 중복 제거 (불필요한 검증 로직 )

Jwts.parser()에는 만료되었는지 확인하는 기능도 포함되어있어서 parser()만 살리고 추가한 기능 삭제

wan2daaa committed:
refact: SpringSecurityConfig 수정 및 기타 설정 추가

- 일반 form 로그인을 진행할 때 경로 permitAll()
-  logout 로직 추가 (Authentication 삭제 및 쿠키 제거
- UserDetailsService 커스텀으로 제작 : findByEmail() 로직으로 작동되게 수정

wan2daaa committed:
feat: Custom UserDetailsService 구현

loadUserByUsername에서 repository에서
email 기준으로 들고오게 진행

wan2daaa committed:
refact: 키워드 위치 변경

wan2daaa committed:
refact: 변수를 메소드에 인라인화 적용

wan2daaa committed:
feat: Exception 과 RestControllerAdvice 생성

wan2daaa committed:
refact: 패키지명이 명확하지 않은것 같아서 패키지 명 변경 및 이동

config/oauth , config/jwt 가 애매한 것 같고, util 안에 있는 것보다 auth 패키지를 만들어 넣는게 조금더 합당하다라고 판단이 되어서 auth패키지 안에 oauth, jwt 패키지 생성

wan2daaa committed:
refact: Exception과 message를 직관적으로 변경

wan2daaa committed:
refact: String 에서 enum에서 들고오게 변경

wan2daaa committed:
refact: TokenAuthenticationFilter accessToken 재발급 로직 추가

accessToken을 검증하고,
-> 검증에 실패하면  refreshToken를 검증한다.
검증에 통과하면,
-> accessToken을 header에 넣어준다

wan2daaa committed:
fix: tokenAuthenticationFilter에 DI 파라미터 추가

5tr1ker reviewed (approved):
고생하셨숩니다!
